packages/affiliates/gas-rebate/VoterGasRebateV2.ts

```diff
@@ -0,0 +1,128 @@
+
+  console.log("Current time:", moment(currentDate).format());
+  console.log("Previous Month Start:", moment(prevMonthStart).format(), "& block", fromBlock);
+  console.log("Previous Month End:", moment(prevMonthEnd).format(), "& block", toBlock);
+
+  // Fetch all commit and reveal events.
+  const voting = await hre.ethers.getContractAt("VotingV2", await getAddress("VotingV2", 1));
+  const commitEvents = await voting.queryFilter(voting.filters.VoteCommitted(), fromBlock, toBlock);
+  const revealEvents = await voting.queryFilter(voting.filters.VoteRevealed(), fromBlock, toBlock);
+
+  // For each event find the associated transaction. We want to refund all transactions that were sent by voters.
+  const transactionsToRefund = await Promise.all(
+    [
+      commitEvents.map(async (commit: any) => voting.provider.getTransactionReceipt(commit.transactionHash)),
+      revealEvents.map(async (reveal: any) => voting.provider.getTransactionReceipt(reveal.transactionHash)),
+    ].flat()
+  );
+
+  console.log(
+    `In aggregate, refund ${commitEvents.length} commits and ${revealEvents.length} Reveals` +
+      ` for a total of ${transactionsToRefund.length} transactions`
+  );
+
+  // Find the associated UMA/ETH price for each transaction. We use the UniswapV3 pool to find the price at the block.
+  // Note that we search for 10000 blocks before to ensure that the range over which we have data spans the entire range.
+  const uniswapPool = await hre.ethers.getContractAt("UniswapV3", "0x157dfa656fdf0d18e1ba94075a53600d81cb3a97");
+  const uniswapPoolEvents = await uniswapPool.queryFilter(uniswapPool.filters.Swap(), fromBlock - 10000, toBlock);
+
+  const shareholderPayoutBN: { [key: string]: BigNumber } = {};
+  // Now, traverse all transactions and calculate the rebate for each.
+  for (const transaction of transactionsToRefund) {
+    // Eth used is the gas used * the gas price.
+    const ethUsed = transaction.gasUsed.mul(transaction.effectiveGasPrice);
+    // Find the nearest UMA/ETH price to the block in which the transaction was included.
+    const associatedUmaEthPrice = findClosestUmaEthPrice(uniswapPoolEvents, transaction.blockNumber);
+    // The rebate is the eth used * the price of 1 ETH in UMA.
+    const resultantRebate = ethUsed.mul(associatedUmaEthPrice).div(ethers.utils.parseUnits("1"));
+    // Save the output to the shareholderPayout object. Append to existing value if it exists.
+    if (!shareholderPayoutBN[transaction.from]) shareholderPayoutBN[transaction.from] = BigNumber.from(0);
+    shareholderPayoutBN[transaction.from] = shareholderPayoutBN[transaction.from].add(resultantRebate);
+  }
+
+  // Create a formatted output that is not bignumbers.
+  const shareholderPayout: { [key: string]: number } = {};
+  for (const [key, value] of Object.entries(shareholderPayoutBN))
+    shareholderPayout[key] = parseFloat(ethers.utils.formatEther(value));
```

Comment:
this cast has a tiny chance to alter number depending on how precise formatEther is, probably just omit it since you dont need math (unless you care if there are quotes or not around the json value)

Reply:
ye, thats ok. we want it to be string.